### PR TITLE
Alerting: Alert list performance improvements

### PR DESCRIPTION
**Pull request by konrad147**
**What this PR does / why we need it**:
This PR improves rendering performance of alert rules list

**Special notes for your reviewer**:
This is the first step to improving the UI performance of the Alerting
I think we still have some space for more improvements

**DEMO**

[AlertListPerf.webm](https://user-images.githubusercontent.com/6293563/193773450-7f925780-d6c8-420b-afd7-3a7da9943cae.webm)

### CODE REVIEW

File: public/app/features/alerting/unified/RuleList.tsx [hunk 1]
```diff
@@ -50,40 +53,45 @@ const RuleList = withErrorBoundary(
...
-        (Object.keys(rulerRuleRequests[name]?.result || {}).length && !rulerRuleRequests[name]?.error)
+
+    const promRequests = Object.entries(promRuleRequests);
+    const allPromLoaded = promRequests.every(
+      ([_, state]) => state.dispatched && (state?.result !== undefined || state?.error !== undefined)
     );
+    const allPromEmpty = promRequests.every(([_, state]) => state.dispatched && state?.result?.length === 0);
+
+    // Trigger data refresh only when the RULE_LIST_POLL_INTERVAL_MS elapsed since the previous load FINISHED
+    const [_, fetchRules] = useAsyncFn(async () => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/RuleList.tsx:70]: 🙌

**Review by gillesdemey — APPROVED** [review]
LGTM, awesome work! 🚀

File: public/app/features/alerting/unified/components/rules/RulesTable.tsx [hunk 2]
```diff
@@ -45,15 +45,9 @@ export const RulesTable: FC<Props> = ({
...
-    const seenKeys: string[] = [];
     return rules.map((rule, ruleIdx) => {
-      let key = JSON.stringify([rule.promRule?.type, rule.labels, rule.query, rule.name, rule.annotations]);
-      if (seenKeys.includes(key)) {
-        key += `-${ruleIdx}`;
-      }
-      seenKeys.push(key);
       return {
-        id: key,
+        id: `${rule.namespace.name}-${rule.group.name}-${rule.name}-${ruleIdx}`,
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/rules/RulesTable.tsx:50]: 🙌💡

File: public/app/features/alerting/unified/hooks/useCombinedRuleNamespaces.ts [hunk 3]
```diff
@@ -129,18 +130,29 @@ function addRulerGroupsToCombinedNamespace(namespace: CombinedRuleNamespace, gro
 }
 
 function addPromGroupsToCombinedNamespace(namespace: CombinedRuleNamespace, groups: RuleGroup[]): void {
+  const existingGroupsByName = new Map<string, CombinedRuleGroup>();
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/hooks/useCombinedRuleNamespaces.ts:133]: 👏

**Review by soniaAguilarPeiron — APPROVED** [review]
That's awesome!! Great job!!! 🚀

**Pull request merged**